### PR TITLE
[HotFix] Logging - One epoch delay on training epoch metrics.

### LINKED ISSUE #4863: trainer.progress_bar_dict values are delayed by one epoch
## 🐛 Bug

`trainer.progress_bar_dict` does not contain what was `self.log`-ed in that epoch's training step, but the previous value.
This also means that there is no value for epoch 0.

### To Reproduce (and to test)

```python
def test_progress_bar_dict_contains_values_on_train_epoch_end(tmpdir):
    class TestModel(BoringModel):
        def training_step(self, *args):
            self.log("foo", torch.tensor(self.current_epoch), on_step=False, on_epoch=True, prog_bar=True)
            return super().training_step(*args)

        def on_train_epoch_end(self, *_):
            self.epoch_end_called = True
            assert self.trainer.progress_bar_dict["foo"] == self.current_epoch

    trainer = Trainer(
        default_root_dir=tmpdir,
        max_epochs=2,
        limit_train_batches=1,
        limit_val_batches=0,
        checkpoint_callback=False,
        logger=False,
        weights_summary=None,
        progress_bar_refresh_rate=0,
    )
    model = TestModel()
    trainer.fit(model)
    assert model.epoch_end_called
```

### Expected behavior

`trainer.progress_bar_dict` contains the values logged in `training_step`

### Environment

**The test fails on master**

The commit that introduced this bug is 9c8701f2e2c09a984ac0322942c8f742a7cc6bc9. The previous one works as expected.

I haven't dived into the changes much since it is a large commit. Maybe the author can help @tchaton

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR update logging logic for training.

Fixes #4863

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
Great!

**Review by SeanNaren — APPROVED** [review]
